### DISCUSSION TITLE
Modularize Reducers with Action-specific functions

### DISCUSSION THREAD

**Discussion by RabugenTom** [pointfreeco/swift-composable-architecture#491] [Ideas]
When composing features, we have to handle child actions in the parent's reducer. This results in most of the cases in `.none` effects without state modification. Exhaustivity of enum cases handling is a double-edged sword and if we pay the guarantee we didn't forget a case with a lot of boilerplate, especially near the root reducer. This guarantee can be open to discussion as serious testing should also prevent this kind of mistake, and we may be tempted to use a `default: return .none` case.

When a child state modifies its state, the parent feature is not automatically notified of the change of its own state which can sometimes propagate to other features, and some effects may need to be produced, or the parent state further adjusted. The solution is to handle explicitly the responsible actions in the parent reducer. When doing this, one must take care that we insert the action-specific derivation above the general case. This kind of burden also composes and we may be forced to write intricate switches. Furthermore, the reducer ends up mixing code related to a lot of different effects. This is not optimal cognitively speaking.

I propose to introduce a higher-order `Reducer` called `when` which corresponds to the handling of a specific Action (a `case ...:` in the source `Reducer`'s switch. It allows to pluck out specific behaviors out of the main body of the reducer. To illustrate the API, imagine a `Child` feature inhering a `String` value from its parent. It can also edit the Parent's value. The parent must react to these changes, and if some specific string `"Secret"` is used, another behavior should be triggered. Here is a vanilla implementation:

```swift
struct ChildState {
  var parentName: String = ""
}

enum ChildAction: Equatable {
  case parentName(String)
  case anotherAction
}

let childReducer = Reducer<ChildState, ChildAction, Void> {
  state, action, _ in
  switch action {
  case let .parentName(name):
    state.parentName = name
    return .none
  case .anotherAction:
    return .none
  }
}


struct ParentState {
  var name: String = "Foo"
  
  var child: ChildState {
    get { .init(parentName: name) }
    set { name = newValue.parentName }
  }
}

enum ParentAction: Equatable {
  case name(String)
  case nameDidChange
  case userKnowsSecret
  case child(ChildAction)
}

let parentReducer = Reducer<ParentState, ParentAction, Void>.combine(
  childReducer.pullback(state: \.child, action: /ParentAction.child, environment: { _ in () }),
  Reducer<ParentState, ParentAction, Void> {
    state, action, _ in
    switch action {
    case let .name(name):
      state.name = name
      return .none
    case .nameDidChange:
      return .fireAndForget { /* … */ }
    case .userKnowsSecret:
      return .fireAndForget { /* … */ }
    //--- Child Actions ---//
    case let .child(.parentName(value)):
      if value == "Secret" {
        return .merge(
          Effect(value: .nameDidChange),
          Effect(value: .userKnowsSecret)
        )
      }
      return Effect(value: .nameDidChange)
    case .child:
      return .none
    }
  }
)
```
Using the `when` higher order reducer, the parent's reducer can be rewritten as:
```swift
// (same as vanilla up to `//--- Child Actions ---//)
let modularParentReducer = Reducer<ParentState, ParentAction, Void>.combine(
  childReducer.pullback(state: \.child, action: /ParentAction.child, environment: { _ in () }),
  Reducer<ParentState, ParentAction, Void> {
    state, action, _ in
    switch action {
    case let .name(name):
      state.name = name
      return .none
    case .nameDidChange:
      return .fireAndForget { /* … */ }
    case .userKnowsSecret:
      return .fireAndForget { /* … */ }
    //--- Child Actions ---//
    case .child:
      return .none
    }
  }
)
.when(/ParentAction.child..ChildAction.parentName) { state, value, environment in
  Effect(value: .nameDidChange)
}
.when(.child(.parentName("Secret"))) { state, environment in
  Effect(value: .userKnowsSecret)
}
```
I've named all functions arguments to show they're available, even if they're not used in this example. In the first derivation, we match `.child(.parentName)` and we emit the "nameDidChange" effect. In the second case, we match `.child(.parentName("Secret"))` explicitly and we emit the "userKnowsSecret" effect. Please note that we didn't need to also emit `nameDidChange` for this specific case as it is already covered by the case above.

In the opposite of the vanilla implementation, we can reorder the two statements without changing the code, and we can comment a whole statement without having to touch anything else. 

I feel this higher-order reducer allows to handle these ad-hoc derivations more neatly. The `CasePath` composition may feel a little awkward, but it could improve in the future or with codegen. This is purely additive and can live outside Core TCA, but I feel it's general enough to be useful for most users. Do you think I should prepare a PR with a case study in this spirit?

Here is the implementation of the `when` reducer. There are three branches: 
- Case with an associated value, handled with `(inout State, Value, Env) -> Effect`
- Case without associated value, handled with `(inout State, Env) -> Effect`
- (Explicit) Equatable match, handled with `(inout State, Env) -> Effect`

```swift
extension Reducer {
  /// Make a `Reducer` return an additional `Effect<Action, Never>` whenever it receives an `Action`'s case
  /// with an associatedValue `Value`. The additional effect is emitted last.
  /// - Parameters:
  ///   - casePath: A CasePath to a case of `Action` with the associatedValue `Value`
  ///   - produce: A function that produce an effect in response to the action.
  func when<Value>(_ casePath: CasePath<Action, Value>,
                   produce: @escaping (inout State, Value, Environment) -> Effect<Action, Never>) -> Self
  {
    combined(with: Self {
      state, action, environment in
      casePath.extract(from: action).map { produce(&state, $0, environment) }
        ?? .none
    })
  }

  /// Make a `Reducer` return an additional `Effect<Action, Never>` whenever it receives an `Action`'s case.
  /// The additional effect is emitted last.
  /// - Parameters:
  ///   - casePath: A CasePath to a case of `Action`
  ///   - produce: A function that produce an effect in response to the action.
  func when(_ casePath: CasePath<Action, Void>,
            produce: @escaping (inout State, Environment) -> Effect<Action, Never>) -> Self
  {
    combined(with: Self {
      state, action, environment in
      casePath.extract(from: action).map { produce(&state, environment) }
        ?? .none
    })
  }

  /// Make a `Reducer` return an additional `Effect<Action, Never>` whenever it receives an equatable `Action`'s case.
  /// This can apply for cases with an explitcit associatedValue like `.setString("Done!")`. The additional effect is emitted last.
  /// - Parameters:
  ///   - casePath: A fully defined case of `Action`.
  ///   - produce: A function that produce an effect in response to the action.
  func when(_ specificCase: Action,
            produce: @escaping (inout State, Environment) -> Effect<Action, Never>) -> Self
    where Action: Equatable
  {
    combined(with: Self {
      state, action, environment in
      guard action == specificCase else { return .none }
      return produce(&state, environment)
    })
  }
}
```

Please let me know what you think!

**Comment by andreyz** [comment]
There's a `.filter` extension on the Reducer in isowords code base that is functionally equivalent. Maybe that'd do the trick?

**Reply by RabugenTom** [reply]
As I'm seeing it, `.filter` restricts an existing reducer effects, which is a "destructive" operation, where the `when` higher order reducer quickly appends new effects to a reducer, which is an "additive" operation. 

In the pitch spirit, the kind of effect you're adding with `when` are living on the parent's level, and are not already present in a child reducer. If it was the case, it would be similar to `.combine(.filter(` indeed, but I don't see any use-case yet in this configuration. So I don't think they'll do the trick here.

**Reply by andreyz** [reply]
> As I'm seeing it, `.filter` restricts an existing reducer effects

I don't think that's true. `.filter` literally filters out **actions** to which the reducer is applied. The reducer itself is immutable.

In your particular case `.when` is functionally equivalent to writing `.init { action, state, environment in _ }.filter { ... } `.

**Reply by RabugenTom** [reply]
I may have expressed myself not clearly. When I say it restricts a reducer effects, it means that you're not mutating your state and not producing side-effects depending on the predicate result. The filtered reducer can produce less effects than the original reducer and the effects it will produce are existing effects from the original reducer. In other words, the range of effects of the filtered reducer is at best the same as the original reducer, most likely smaller. That was the meaning of "destructive" that I put between quotes. I should have use "substractive".

The `when` reducer allows to add new effects, possibly never written elsewhere in any other reducer. I'm not seeing how to the modular reducer of the example can produce `Effect(value: .userKnowsSecret)` when the user enters "Secret", by using filtering only and without having to write down this logic in the main reducer itself, like in the vanilla case, but I haven't used `filter` myself, so I may be wrong.

**Comment by lukeredpath** [comment]
Could you achieve something similar using the `onChange` reducer shown here:

https://github.com/pointfreeco/isowords/blob/main/Sources/TcaHelpers/OnChange.swift

It seems the advantage here is that you can more accurately express the idea of "do something when this state changes" without having to care about which action triggered that change.

**Reply by RabugenTom** [reply]
Yes, I've seen it in latest pointfree episode. It indeed covers nicely the state changes case which was presented in the pitch.
However, `when` reducer also works for "Notification"-like pure actions, which can be useful on its own I guess.

**Reply by RabugenTom** [reply]
After reflection, `onChange`, which seems perfectly fine in many cases, inherits the problems of bidirectional architectures like vanilla SwiftUI: your value changed, but you don't know the reason why. It is perfectly fine in most of the cases, but you may want to produce different effects in some cases. For example, when your value change because the user has edited it and you need to save it, or when it changes because you just received a new value from Core Data sync and you shouldn't produce any side-effect. Both cases will be spotted by `onChange`.

`onChange` is not problematic but It bears some logic in it, and that delimits its own class of legitimate uses, which I think doesn't completely overlaps `when` uses for state mutation.

`when` is merely syntactic sugar to decompose action matching. It's only a way to pluck out an action to produce an effect in response, whereas the action mutates the state, or is a mere notification. As such, it doesn't need to access private properties of TCA types and can be implemented as a third party extension. Maybe I'll wait for the Community TCA repo.

**Reply by andreyz** [reply]
Actually onChange also provides an action that caused the change. Check out the signature and it's being used with an action specific code in isowords (AppView.Swift)

**Reply by RabugenTom** [reply]
You're right!
`onChange` will do the trick most of the time then.
